src/librustc_trans/trans/callee.rs

```diff
@@ -765,8 +765,16 @@ pub fn trans_call_inner<'a, 'blk, 'tcx, F>(bcx: Block<'blk, 'tcx>,
     if is_rust_fn {
         let mut llargs = Vec::new();
 
-        if let (ty::FnConverging(ret_ty), Some(llretslot)) = (ret_ty, opt_llretslot) {
+        if let (ty::FnConverging(ret_ty), Some(mut llretslot)) = (ret_ty, opt_llretslot) {
             if type_of::return_uses_outptr(ccx, ret_ty) {
+                let llformal_ret_ty = type_of::type_of(ccx, ret_ty).ptr_to();
+                let llret_ty = common::val_ty(llretslot);
+                if llformal_ret_ty != llret_ty {
+                    // this could happen due to e.g. subtyping
```

Comment:
We kind of assertion do you have in mind? We just have the LLVM type of the retslot here (can you lookup the original Ty from that?), I don't think we can do much with that, can we?

Ultimately, we should probably ignore the trait bounds when creating the LLVM types, but this seemed like the obvious quick fix because `trans_arg_datum` already does the same.

And once LLVM drops typed pointers, this is all moot anyway.